bigtable/client/row_range.cc

```diff
@@ -0,0 +1,446 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "bigtable/client/row_range.h"
+
+#include <iostream>
+#include <thread>
+
+#include <absl/memory/memory.h>
+#include <absl/strings/str_cat.h>
+
+namespace btproto = ::google::bigtable::v2;
+
+namespace bigtable {
+inline namespace BIGTABLE_CLIENT_NS {
+const std::string RowRange::SAMPLE_KEY = "";
+;
+
+/**
+ * The contructor for RowRange: [begin, end).
+ * In the constructor and all the builder methods, We don't check
+ * if begin <= end. This anomaly is caught during range manipulation
+ * (i.e., intersect, contains)
+ */
+RowRange::RowRange(std::string begin, std::string end) {
+  range_.set_start_key_closed(std::move(begin));
+  if (!end.empty()) range_.set_end_key_open(std::move(end));
+}
+
+/**
+ * An empty range satisfies the following conditions:
+ *   - Any semi-open or open range [x, x), (x,x], (x,x) where x could be empty.
+ *   - 'end' is set (otherwise it would be an infinite range).
+ *   - Not a closed interval [x,x] with non-empty x.
+ *
+ * We use the right-open [x,x) as the default empty range.
+ * For convenience and the readability of unit tests,
```

Comment:
Convenient, maybe, more readable?  Not for me.

Reply:
Please refer to the first reply above. Thanks.